src/v/cloud_storage/inventory/inv_ops.h

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_storage/inventory/aws_ops.h"
+
+namespace cloud_storage::inventory {
+
+using ops_t = inv_ops_variant<aws_ops>;
```

Comment:
nit: might be a bit premature, but it feels like it'd be simpler to actually use this as a base class and use polymorphism. Or perhaps added to the cloud_storage_clients/client.h interface? That would feel a bit more natural, since it's a clear division between the different platforms that can be tested without cross dependencies

Reply:
>  it'd be simpler to actually use this as a base class and use polymorphism

I opted to use sum type here as the final set of implementations is a closed set of three implementations. Do you see issues with complexity here that could be simplified?

> Or perhaps added to the cloud_storage_clients/client.h interface? That would feel a bit more natural, since it's a clear division between the different platforms that can be tested without cross dependencies

Do you mean that the implementation should be moved down to `cloud_storage_clients`? I briefly mentioned this in the cover letter but to elaborate, there are many differences between AWS and GCP implementations for inventory operations in terms of responses, request structure and even figuring out what the latest report is, which will require creating a GCP specific client. 

We do not need a separate GCP client so far because the `cloud_storage_clients` module works on low level HTTP calls and is able to keep GCP and S3 under the same interface because of that.

The inventory related operations are higher level so it makes sense to me to implement them here rather than as more primitive HTTP calls. WDYT?

Reply:
>issues with complexity here that could be simplified?

I guess it'll depend on how the rest of the implementation goes. My comment is more of a knee-jerk reaction: maybe there will be less platform-specific bits in there than I'm imagining, but without knowing more about how each platform will be configured, stronger separation feels natural since that's what we do between s3/abs.

That said, I can see how it'd be much simpler to use `remote` as the primitive instead of baking this deeper into something like the client, and I don't think variant vs polymorphism will change much given we're relying on such a high level primitive.

Good point about AWS vs GCP, though even that seems like a leaky abstraction that will eventually be ironed out and separated. I don't feel strongly that this should happen now though.

Reply:
These are good points, I agree that we should implement things lower down in remote/cloud_storage_clients where possible, but I think these operations belong in a higher level API. 

Right now we have the layers of implementation as:

* cloud_storage_clients (s3/abs) : lowest layer - usually methods map directly to HTTP verbs. S3 and GCP are assumed to be the same
* cloud_storage::remote: usually methods map to redpanda cloud storage object related actions, with some generic calls, eg segment operations, manifest operations. Provides retry, client pool etc.
* inventory (as seen/planned here): methods map to inventory specific operations, will often contain sequences of multiple calls into remote and will need to manage the state of calls. eg
  * AWS download report: check if a checksum file is present, then downloads a CSV file from a computed path, might need downloading a manifest and parsing it first. https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-inventory-location.html
  * GCP download report: get the UUID for redpanda inventory config, maybe cache it in memory, check if that UUID's checksum is present, then download a JSON manifest, parse out the actual inventory file paths from it, and download those. These report file paths may be multiple as GCP divides report into shards, so handle returning the output stream from the multiple shards. https://cloud.google.com/storage/docs/insights/using-inventory-reports#download-reports
  * Azure: a recursive listing might be required to find the latest checksum and inventory, as reports are kept in a path which contains date and time components. https://learn.microsoft.com/en-us/azure/storage/blobs/blob-inventory#inventory-output

The inventory API methods will provide an interface to hide these platform specific details and allow the caller to just call eg `download_latest_report` and get a stream to use, and hide the successive calls required to do these actions.

It is certainly possible to put these into remote, but I think that it is easier to understand by keeping the inventory related methods secluded as these are (IMO) higher level operations. 

> maybe there will be less platform-specific bits in there than I'm imagining, but without knowing more about how each platform will be configured, stronger separation feels natural since that's what we do between s3/abs.

There is quite a bit of platform/vendor specific differences in terms of API calls, the plan is to let `aws_ops`, `gcs_ops` and `azure_ops` (the last two to be implemented later) implement the platform specific details, and `inv_ops` just uses `visit` to proxy the calls to the underlying variant. 

So the call chain looks like:

```
// AWS
inv_ops (dl_report) -> aws_ops (dl_report) ->
  call 1 -> remote -> s3_client
  call 2 -> remote -> s3_client
  ...
//GCP
inv_ops (dl_report) -> gcs_ops (dl_report) ->
  call 1 -> remote -> s3_client
  call 2 -> remote -> s3_client
  call 3 -> remote -> s3_client
  call 4 -> remote -> s3_client
...
// Azure
...
```

Reply:
My reasoning is that when writing a utility such as a library which may be freely extended by classes which are not known at the time of the implementation of the base class, runtime polymorphism (interface based design) is a better fit. 

Here we know upfront that there are only going to be three implementations, a sum type makes more sense to me. 

In other words, saying that this object can only ever be one of vendor_api_a, vendor_api_b or vendor_api_c seems a better fit to me than saying that this object is a vendor_api which fulfills the expected contract. 

The contract is still enforced by the `vendor_ops_provider` concept.

> otherwise we end up generating extra code

Do you mean extra code generation by using the variant, or because of the templates? The template definition is only for the contracts and is fairly isolated.

Reply:
I'm open to suggestions though, as I understand sum types have better ergonomics in some languages than others and `std::variant` is a recent addition, if there are any major concerns with using `std::variant` here, the design is not tied to it and we can make changes. 

I still think the variant+visit based approach looks easier to understand and more concise than PIMPL.